
# Acquiring data for the last 10 years using longitude & latitude values for shanghai
## Temperature,	Wind Direction,	Wind Speed,	Humidity & Pressure

In [ ]:
!pip install meteostat

In [ ]:
from meteostat import Point, Hourly
import pandas as pd
from datetime import datetime, timedelta

# Define the location for Shanghai
shanghai = Point(31.2304, 121.4737)

# Specify the date range for the past 30 days
end_date = datetime.now()
start_date = pd.to_datetime('2014-01-01')

# Fetch hourly historical data
data = Hourly(shanghai, start=start_date, end=end_date)

# Convert data to a DataFrame
df = data.fetch()

# Display the DataFrame
print(df.head())

# Optionally, save to CSV
df.to_csv('shanghai_weather_data.csv')

In [40]:
data=pd.read_csv('shanghai_weather_data.csv')
data.head()

time  temp  dwpt  rhum  prcp  snow   wdir  wspd  wpgt  pres  \
0  2014-01-01 00:00:00   4.0  -4.0  56.0   NaN   NaN  230.0   7.2   NaN   NaN   
1  2014-01-01 01:00:00   7.0  -4.1  45.0   NaN   NaN  230.0  10.8   NaN   NaN   
2  2014-01-01 02:00:00  10.0  -5.2  34.0   NaN   NaN  240.0  14.4   NaN   NaN   
3  2014-01-01 03:00:00  13.0  -5.1  28.0   NaN   NaN  260.0  14.4   NaN   NaN   
4  2014-01-01 04:00:00  14.0  -6.8  23.0   NaN   NaN  250.0  18.0   NaN   NaN   

   tsun  coco  
0   NaN   NaN  
1   NaN   NaN  
2   NaN   NaN  
3   NaN   NaN  
4   NaN   NaN

## filtering needed columns and renaming them for easy understanding

In [5]:
data_filtered = data[['time', 'temp', 'wdir', 'wspd', 'rhum', 'pres']].rename(columns={
    'time': 'Timestamp',
    'temp': 'Temperature',
    'wdir': 'Wind Direction',
    'wspd': 'Wind Speed',
    'rhum': 'Humidity',
    'pres': 'Pressure'
})


In [6]:
data_filtered.isnull().sum()

Timestamp             0
Temperature          62
Wind Direction     2766
Wind Speed          171
Humidity             62
Pressure          14960
dtype: int64

In [7]:
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96064 entries, 0 to 96063
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Timestamp       96064 non-null  object 
 1   Temperature     96002 non-null  float64
 2   Wind Direction  93298 non-null  float64
 3   Wind Speed      95893 non-null  float64
 4   Humidity        96002 non-null  float64
 5   Pressure        81104 non-null  float64
dtypes: float64(5), object(1)
memory usage: 4.4+ MB


In [8]:
import pandas as pd

# Assuming you already have your DataFrame as 'df_hourly'
# Calculate percentage of missing values
missing_percentages = data_filtered.isnull().mean() * 100

# Filter for columns with missing data
missing_data_columns = missing_percentages[missing_percentages > 0]

# Display the results
print("Percentage of missing values in columns with missing data:")
print(missing_data_columns)

Percentage of missing values in columns with missing data:
Temperature        0.064540
Wind Direction     2.879330
Wind Speed         0.178006
Humidity           0.064540
Pressure          15.572951
dtype: float64


# Checking rows with missing values to find a way for imputation

In [9]:
null_wind_speed=data_filtered[data_filtered['Wind Speed'].isnull()]
null_wind_speed.head(50)


Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
190    2014-01-08 22:00:00          NaN             NaN         NaN       NaN   
290    2014-01-13 02:00:00          NaN             NaN         NaN       NaN   
329    2014-01-14 17:00:00          NaN             NaN         NaN       NaN   
1062   2014-02-14 06:00:00          NaN             NaN         NaN       NaN   
1091   2014-02-15 11:00:00          NaN             NaN         NaN       NaN   
1092   2014-02-15 12:00:00          NaN             NaN         NaN       NaN   
1166   2014-02-18 14:00:00          NaN             NaN         NaN       NaN   
3014   2014-05-06 14:00:00          NaN             NaN         NaN       NaN   
3015   2014-05-06 15:00:00          NaN             NaN         NaN       NaN   
3038   2014-05-07 14:00:00          NaN             NaN         NaN       NaN   
3139   2014-05-11 19:00:00          NaN             NaN         NaN       NaN   
3140   2014-05-11 20:00:00          NaN             NaN         NaN       NaN   
3141   2014-05-11 21:00:00          NaN             NaN         NaN       NaN   
3307   2014-05-18 19:00:00          NaN             NaN         NaN       NaN   
3903   2014-06-12 15:00:00          NaN             NaN         NaN       NaN   
3908   2014-06-12 20:00:00          NaN             NaN         NaN       NaN   
3933   2014-06-13 21:00:00          NaN             NaN         NaN       NaN   
4501   2014-07-07 13:00:00          NaN             NaN         NaN       NaN   
4502   2014-07-07 14:00:00          NaN             NaN         NaN       NaN   
4503   2014-07-07 15:00:00          NaN             NaN         NaN       NaN   
4621   2014-07-12 13:00:00          NaN             NaN         NaN       NaN   
4622   2014-07-12 14:00:00          NaN             NaN         NaN       NaN   
4623   2014-07-12 15:00:00          NaN             NaN         NaN       NaN   
4624   2014-07-12 16:00:00          NaN             NaN         NaN       NaN   
4625   2014-07-12 17:00:00          NaN             NaN         NaN       NaN   
4626   2014-07-12 18:00:00          NaN             NaN         NaN       NaN   
4627   2014-07-12 19:00:00          NaN             NaN         NaN       NaN   
4628   2014-07-12 20:00:00          NaN             NaN         NaN       NaN   
4629   2014-07-12 21:00:00          NaN             NaN         NaN       NaN   
4630   2014-07-12 22:00:00          NaN             NaN         NaN       NaN   
4632   2014-07-13 00:00:00          NaN             NaN         NaN       NaN   
4635   2014-07-13 03:00:00          NaN             NaN         NaN       NaN   
4641   2014-07-13 09:00:00          NaN             NaN         NaN       NaN   
5149   2014-08-03 13:00:00          NaN             NaN         NaN       NaN   
5902   2014-09-03 22:00:00          NaN             NaN         NaN       NaN   
6256   2014-09-18 16:00:00          NaN             NaN         NaN       NaN   
6351   2014-09-22 15:00:00          NaN             NaN         NaN       NaN   
6543   2014-09-30 15:00:00          NaN             NaN         NaN       NaN   
6832   2014-10-12 16:00:00          NaN             NaN         NaN       NaN   
6833   2014-10-12 17:00:00          NaN             NaN         NaN       NaN   
6834   2014-10-12 18:00:00          NaN             NaN         NaN       NaN   
6835   2014-10-12 19:00:00          NaN             NaN         NaN       NaN   
7864   2014-11-24 16:00:00          NaN             NaN         NaN       NaN   
7911   2014-11-26 15:00:00          NaN             NaN         NaN       NaN   
8252   2014-12-10 20:00:00          NaN             NaN         NaN       NaN   
8255   2014-12-10 23:00:00          NaN             NaN         NaN       NaN   
8331   2014-12-14 03:00:00          NaN             NaN         NaN       NaN   
10950  2015-04-02 06:00:00          NaN             NaN         NaN       NaN   
11094  2015-04-08 06:00:00          NaN         

In [10]:
null_temp=data_filtered[(data_filtered['Temperature'].isnull())]
null_temp.head(50)

Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
190    2014-01-08 22:00:00          NaN             NaN         NaN       NaN   
290    2014-01-13 02:00:00          NaN             NaN         NaN       NaN   
329    2014-01-14 17:00:00          NaN             NaN         NaN       NaN   
1062   2014-02-14 06:00:00          NaN             NaN         NaN       NaN   
1091   2014-02-15 11:00:00          NaN             NaN         NaN       NaN   
1092   2014-02-15 12:00:00          NaN             NaN         NaN       NaN   
1166   2014-02-18 14:00:00          NaN             NaN         NaN       NaN   
3014   2014-05-06 14:00:00          NaN             NaN         NaN       NaN   
3015   2014-05-06 15:00:00          NaN             NaN         NaN       NaN   
3038   2014-05-07 14:00:00          NaN             NaN         NaN       NaN   
3139   2014-05-11 19:00:00          NaN             NaN         NaN       NaN   
3140   2014-05-11 20:00:00          NaN             NaN         NaN       NaN   
3141   2014-05-11 21:00:00          NaN             NaN         NaN       NaN   
3307   2014-05-18 19:00:00          NaN             NaN         NaN       NaN   
3903   2014-06-12 15:00:00          NaN             NaN         NaN       NaN   
3908   2014-06-12 20:00:00          NaN             NaN         NaN       NaN   
3933   2014-06-13 21:00:00          NaN             NaN         NaN       NaN   
4501   2014-07-07 13:00:00          NaN             NaN         NaN       NaN   
4502   2014-07-07 14:00:00          NaN             NaN         NaN       NaN   
4503   2014-07-07 15:00:00          NaN             NaN         NaN       NaN   
4621   2014-07-12 13:00:00          NaN             NaN         NaN       NaN   
4622   2014-07-12 14:00:00          NaN             NaN         NaN       NaN   
4623   2014-07-12 15:00:00          NaN             NaN         NaN       NaN   
4624   2014-07-12 16:00:00          NaN             NaN         NaN       NaN   
4625   2014-07-12 17:00:00          NaN             NaN         NaN       NaN   
4626   2014-07-12 18:00:00          NaN             NaN         NaN       NaN   
4627   2014-07-12 19:00:00          NaN             NaN         NaN       NaN   
4628   2014-07-12 20:00:00          NaN             NaN         NaN       NaN   
4629   2014-07-12 21:00:00          NaN             NaN         NaN       NaN   
4630   2014-07-12 22:00:00          NaN             NaN         NaN       NaN   
4632   2014-07-13 00:00:00          NaN             NaN         NaN       NaN   
4635   2014-07-13 03:00:00          NaN             NaN         NaN       NaN   
4641   2014-07-13 09:00:00          NaN             NaN         NaN       NaN   
5149   2014-08-03 13:00:00          NaN             NaN         NaN       NaN   
5902   2014-09-03 22:00:00          NaN             NaN         NaN       NaN   
6256   2014-09-18 16:00:00          NaN             NaN         NaN       NaN   
6351   2014-09-22 15:00:00          NaN             NaN         NaN       NaN   
6543   2014-09-30 15:00:00          NaN             NaN         NaN       NaN   
6832   2014-10-12 16:00:00          NaN             NaN         NaN       NaN   
6833   2014-10-12 17:00:00          NaN             NaN         NaN       NaN   
6834   2014-10-12 18:00:00          NaN             NaN         NaN       NaN   
6835   2014-10-12 19:00:00          NaN             NaN         NaN       NaN   
7864   2014-11-24 16:00:00          NaN             NaN         NaN       NaN   
7911   2014-11-26 15:00:00          NaN             NaN         NaN       NaN   
8252   2014-12-10 20:00:00          NaN             NaN         NaN       NaN   
8255   2014-12-10 23:00:00          NaN             NaN         NaN       NaN   
8331   2014-12-14 03:00:00          NaN             NaN         NaN       NaN   
8889   2015-01-06 09:00:00          NaN            20.0        21.6       NaN   
10137  2015-02-27 09:00:00          NaN         

In [11]:
print(data_filtered.iloc[180:200,:])

               Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
180  2014-01-08 12:00:00          5.0           360.0        25.2      75.0   
181  2014-01-08 13:00:00          4.0           360.0        25.2      75.0   
182  2014-01-08 14:00:00          3.0           360.0        14.4      75.0   
183  2014-01-08 15:00:00          3.0           350.0        14.4      75.0   
184  2014-01-08 16:00:00          3.0           350.0        14.4      75.0   
185  2014-01-08 17:00:00          3.0           350.0        18.0      70.0   
186  2014-01-08 18:00:00          2.0           360.0        21.6      75.0   
187  2014-01-08 19:00:00          2.0           350.0        14.4      69.0   
188  2014-01-08 20:00:00          2.0           360.0        14.4      75.0   
189  2014-01-08 21:00:00          1.0           350.0        10.8      75.0   
190  2014-01-08 22:00:00          NaN             NaN         NaN       NaN   
191  2014-01-08 23:00:00          1.0           330.

# Above EDA shows missing values with small and bigger gaps in data with respect to time.
## We impute gaps with upto 2 missing value using interpolation
### i)for gradual factors like temperature we use linear interpolation 
### ii)quadratic for Humidity to capture some curve in the data pattern.
### iii)cubic for Wind_Speed to better capture potential fluctuations.
### iv)nearest for Pressure since this variable may not change drastically in the short term.




In [ ]:
!pip install scipy

In [14]:
# Step 1: Interpolate small gaps (up to 2 consecutive missing values) with specified methods
data_filtered['Temperature'] = data_filtered['Temperature'].interpolate(method='linear', limit=2)
data_filtered['Humidity'] = data_filtered['Humidity'].interpolate(method='quadratic', limit=2)
data_filtered['Wind Speed'] = data_filtered['Wind Speed'].interpolate(method='cubic', limit=2)
data_filtered['Pressure'] = data_filtered['Pressure'].interpolate(method='nearest', limit=2)

## Using a 7-day rolling average to fill larger gaps, which helps capture broader patterns and minimizes the effect of missing data.

In [15]:
data_filtered['Temperature'] = data_filtered['Temperature'].fillna(data_filtered['Temperature'].rolling(window=24*7, min_periods=1).mean())
data_filtered['Humidity'] = data_filtered['Humidity'].fillna(data_filtered['Humidity'].rolling(window=24*7, min_periods=1).mean())
data_filtered['Wind Speed'] = data_filtered['Wind Speed'].fillna(data_filtered['Wind Speed'].rolling(window=24*7, min_periods=1).mean())
data_filtered['Pressure'] = data_filtered['Pressure'].fillna(data_filtered['Pressure'].rolling(window=24*7, min_periods=1).mean())

## Checking for null values after imputation

In [16]:
data_filtered.isnull().sum()

Timestamp             0
Temperature           0
Wind Direction     2766
Wind Speed            0
Humidity              0
Pressure          14010
dtype: int64

## presure still has missing values

In [17]:
pressure_null=data_filtered[(data_filtered['Pressure'].isnull())]
pressure_null.head(50)

Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
0   2014-01-01 00:00:00          4.0           230.0         7.2      56.0   
1   2014-01-01 01:00:00          7.0           230.0        10.8      45.0   
2   2014-01-01 02:00:00         10.0           240.0        14.4      34.0   
3   2014-01-01 03:00:00         13.0           260.0        14.4      28.0   
4   2014-01-01 04:00:00         14.0           250.0        18.0      23.0   
5   2014-01-01 05:00:00         16.0           260.0        18.0      17.0   
6   2014-01-01 06:00:00         16.0           250.0        18.0      18.0   
7   2014-01-01 07:00:00         16.0           270.0        18.0      15.0   
8   2014-01-01 08:00:00         15.0           280.0        14.4      18.0   
9   2014-01-01 09:00:00         13.0           250.0         7.2      22.0   
10  2014-01-01 10:00:00         12.0             NaN         3.6      24.0   
11  2014-01-01 11:00:00         10.0             NaN         0.0      30.0   
12  2014-01-01 12:00:00         10.0           170.0         7.2      32.0   
13  2014-01-01 13:00:00          9.0             NaN         3.6      37.0   
14  2014-01-01 14:00:00          8.0             NaN         3.6      39.0   
15  2014-01-01 15:00:00          7.0             NaN         0.0      42.0   
16  2014-01-01 16:00:00          5.0           170.0         7.2      52.0   
17  2014-01-01 17:00:00          5.0             NaN         0.0      56.0   
18  2014-01-01 18:00:00          3.0             NaN         0.0      60.0   
19  2014-01-01 19:00:00          3.0             NaN         0.0      65.0   
20  2014-01-01 20:00:00          2.0             NaN         0.0      75.0   
21  2014-01-01 21:00:00          2.0             NaN         0.0      69.0   
22  2014-01-01 22:00:00          2.0             NaN         0.0      69.0   
23  2014-01-01 23:00:00          2.0             NaN         0.0      69.0   
24  2014-01-02 00:00:00          6.0            70.0         3.6      65.0   
25  2014-01-02 01:00:00         11.0           130.0        10.8      47.0   
26  2014-01-02 02:00:00         13.0           130.0        14.4      33.0   
27  2014-01-02 03:00:00         15.0           110.0        14.4      29.0   
28  2014-01-02 04:00:00         16.0           120.0        14.4      23.0   
29  2014-01-02 05:00:00         17.0           120.0        14.4      16.0   
30  2014-01-02 06:00:00         18.0           150.0        18.0      18.0   
31  2014-01-02 07:00:00         17.0           150.0        21.6      22.0   
32  2014-01-02 08:00:00         16.0           150.0        18.0      31.0   
33  2014-01-02 09:00:00         15.0           130.0        18.0      39.0   
34  2014-01-02 10:00:00         13.0           120.0        18.0      51.0   
35  2014-01-02 11:00:00         12.0           120.0        14.4      58.0   
36  2014-01-02 12:00:00         12.0           120.0        14.4      62.0   
37  2014-01-02 13:00:00         11.0           130.0        14.4      67.0   
38  2014-01-02 14:00:00         10.0           130.0        10.8      71.0   
39  2014-01-02 15:00:00         11.0           140.0        14.4      71.0   
40  2014-01-02 16:00:00         11.0           150.0        14.4      71.0   
41  2014-01-02 17:00:00         11.0             NaN         3.6      71.0   
42  2014-01-02 18:00:00         10.0           160.0         3.6      76.0   
43  2014-01-02 19:00:00          9.0             NaN         0.0      81.0   
44  2014-01-02 20:00:00          9.0             NaN         0.0      81.0   
45  2014-01-02 21:00:00          9.0             NaN         0.0      81.0   
46  2014-01-02 22:00:00          9.0           340.0         3.6      81.0   
47  2014-01-02 23:00:00          9.0           280.0         3.6      81.0   
48  2014-01-03 00:00:00         10.0             NaN         3.6      76.0   
49  2014-01-03 01:00:00         11.0           300.0        14.4      71.0   

    Pressure  
0        NaN  
1    

In [18]:
variance = data_filtered['Pressure'].var()
std_dev = data_filtered['Pressure'].std()
print("variance:",variance, "standard deviation:", std_dev)

variance: 95.3852805969902 standard deviation: 9.766538823810112


In [19]:
print(data_filtered.iloc[49:100,:])

              Timestamp  Temperature  Wind Direction  Wind Speed  Humidity  \
49  2014-01-03 01:00:00         11.0           300.0        14.4      71.0   
50  2014-01-03 02:00:00         12.0           320.0        18.0      62.0   
51  2014-01-03 03:00:00         13.0           320.0        18.0      54.0   
52  2014-01-03 04:00:00         15.0           320.0        21.6      41.0   
53  2014-01-03 05:00:00         16.0           330.0        18.0      27.0   
54  2014-01-03 06:00:00         17.0           340.0        21.6      22.0   
55  2014-01-03 07:00:00         16.0           340.0        18.0      18.0   
56  2014-01-03 08:00:00         15.0           330.0        18.0      20.0   
57  2014-01-03 09:00:00         14.0           320.0        14.4      23.0   
58  2014-01-03 10:00:00         12.0           340.0        10.8      28.0   
59  2014-01-03 11:00:00         12.0           320.0        10.8      30.0   
60  2014-01-03 12:00:00         12.0           280.0        10.8

In [41]:
pressure_notnull=data_filtered[(data_filtered['Pressure'].notnull())]
pressure_notnull.head()             

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2014-01-01 00:00:00          4.0           230.0         7.2      56.0   
2014-01-01 01:00:00          7.0           230.0        10.8      45.0   
2014-01-01 02:00:00         10.0           240.0        14.4      34.0   
2014-01-01 03:00:00         13.0           260.0        14.4      28.0   
2014-01-01 04:00:00         14.0           250.0        18.0      23.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2014-01-01 00:00:00    1026.0 -0.766044 -0.642788                        230.0  
2014-01-01 01:00:00    1026.0 -0.766044 -0.642788                        230.0  
2014-01-01 02:00:00    1026.0 -0.866025 -0.500000                        240.0  
2014-01-01 03:00:00    1026.0 -0.984808 -0.173648                        260.0  
2014-01-01 04:00:00    1026.0 -0.939693 -0.342020                        250.0

## pressure doesnot vary frequently and has large gaps so we impute the missing value with median/mean of that specific month 

In [21]:
data_filtered['Timestamp'] = pd.to_datetime(data_filtered['Timestamp'])

# Fill missing values in 'Pressure' based on monthly median
data_filtered['Pressure'] = data_filtered['Pressure'].fillna(
    data_filtered['Pressure'].groupby(data_filtered['Timestamp'].dt.month).transform('median')
)


## only wind direction has null values now

In [22]:
data_filtered.isnull().sum()

Timestamp            0
Temperature          0
Wind Direction    2766
Wind Speed           0
Humidity             0
Pressure             0
dtype: int64

In [42]:
data_filtered.tail(10)

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2024-12-16 06:00:00          9.1           262.0         7.4      31.0   
2024-12-16 07:00:00          8.9           264.0         7.4      33.0   
2024-12-16 08:00:00          8.8           267.0         5.5      35.0   
2024-12-16 09:00:00          8.7           271.0         5.5      37.0   
2024-12-16 10:00:00          8.1           251.0         5.5      40.0   
2024-12-16 11:00:00          7.6           229.0         3.7      43.0   
2024-12-16 12:00:00          7.0           210.0         3.7      47.0   
2024-12-16 13:00:00          6.6           211.0         3.7      50.0   
2024-12-16 14:00:00          6.2           213.0         3.7      54.0   
2024-12-16 15:00:00          5.8           215.0         5.5      59.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2024-12-16 06:00:00    1024.8 -0.990268 -0.139173                        262.0  
2024-12-16 07:00:00    1024.8 -0.994522 -0.104528                        264.0  
2024-12-16 08:00:00    1024.9 -0.998630 -0.052336                        267.0  
2024-12-16 09:00:00    1024.9 -0.999848  0.017452                        271.0  
2024-12-16 10:00:00    1025.2 -0.945519 -0.325568                        251.0  
2024-12-16 11:00:00    1025.5 -0.754710 -0.656059                        229.0  
2024-12-16 12:00:00    1025.7 -0.500000 -0.866025                        210.0  
2024-12-16 13:00:00    1025.9 -0.515038 -0.857167                        211.0  
2024-12-16 14:00:00    1026.0 -0.544639 -0.838671                        213.0  
2024-12-16 15:00:00    1026.1 -0.573576 -0.819152                        215.0

### Filtering rows with 0 wind speed and corresponding wind direction

In [24]:
# 0 wind speed but wind direction values available
zero_speed_direction= data_filtered[(data_filtered['Wind Speed'] == 0) & (data_filtered['Wind Direction'].notnull())]
#there are 559 rows with no wind speed but direction mentioned
zero_speed_direction[['Wind Speed', 'Wind Direction']].describe()


Wind Speed  Wind Direction
count       572.0      572.000000
mean          0.0       53.590909
std           0.0      102.406236
min           0.0        0.000000
25%           0.0        0.000000
50%           0.0        0.000000
75%           0.0       61.250000
max           0.0      358.000000

## Filtering rows with null wind direction
### many instances with wind speed equal to 0 and less wind speeds 

In [43]:
null_direction=data_filtered[data_filtered['Wind Direction'].isnull()]
null_direction.head(10)

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2014-01-01 15:00:00          7.0             NaN         0.0      42.0   
2014-01-01 19:00:00          3.0             NaN         0.0      65.0   
2014-01-01 20:00:00          2.0             NaN         0.0      75.0   
2014-01-01 21:00:00          2.0             NaN         0.0      69.0   
2014-01-01 22:00:00          2.0             NaN         0.0      69.0   
2014-01-01 23:00:00          2.0             NaN         0.0      69.0   
2014-01-02 21:00:00          9.0             NaN         0.0      81.0   
2014-01-05 05:00:00         10.0             NaN         3.6      54.0   
2014-01-05 06:00:00         11.0             NaN         3.6      47.0   
2014-01-05 07:00:00         11.0             NaN         3.6      47.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2014-01-01 15:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 19:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 20:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 21:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 22:00:00    1026.0       NaN       NaN                          NaN  
2014-01-01 23:00:00    1026.0       NaN       NaN                          NaN  
2014-01-02 21:00:00    1026.0       NaN       NaN                          NaN  
2014-01-05 05:00:00    1026.0       NaN       NaN                          NaN  
2014-01-05 06:00:00    1026.0       NaN       NaN                          NaN  
2014-01-05 07:00:00    1026.0       NaN       NaN                          NaN

## examining rows with wind direction values present to help with imputing missing values

In [44]:
wind_direction_not_null= data_filtered[data_filtered['Wind Direction'].notnull()]
wind_direction_not_null.tail(20)

Temperature  Wind Direction  Wind Speed  Humidity  \
Timestamp                                                                
2024-12-15 20:00:00          1.0           230.0         7.0      69.0   
2024-12-15 21:00:00          1.0           250.0         4.0      69.0   
2024-12-15 22:00:00          0.0           210.0         7.0      75.0   
2024-12-15 23:00:00          2.0           250.0         4.0      70.0   
2024-12-16 00:00:00          2.2           255.0         3.7      62.0   
2024-12-16 01:00:00          7.0           260.0         7.0      46.0   
2024-12-16 02:00:00          9.0           280.0        11.0      37.0   
2024-12-16 03:00:00         10.0           250.0        14.0      37.0   
2024-12-16 04:00:00          8.5           248.0         7.4      34.0   
2024-12-16 05:00:00          8.8           256.0         7.4      33.0   
2024-12-16 06:00:00          9.1           262.0         7.4      31.0   
2024-12-16 07:00:00          8.9           264.0         7.4      33.0   
2024-12-16 08:00:00          8.8           267.0         5.5      35.0   
2024-12-16 09:00:00          8.7           271.0         5.5      37.0   
2024-12-16 10:00:00          8.1           251.0         5.5      40.0   
2024-12-16 11:00:00          7.6           229.0         3.7      43.0   
2024-12-16 12:00:00          7.0           210.0         3.7      47.0   
2024-12-16 13:00:00          6.6           211.0         3.7      50.0   
2024-12-16 14:00:00          6.2           213.0         3.7      54.0   
2024-12-16 15:00:00          5.8           215.0         5.5      59.0   

                     Pressure  wdir_sin  wdir_cos  Wind_Direction_Interpolated  
Timestamp                                                                       
2024-12-15 20:00:00    1027.0 -0.766044 -0.642788                        230.0  
2024-12-15 21:00:00    1027.0 -0.939693 -0.342020                        250.0  
2024-12-15 22:00:00    1028.0 -0.500000 -0.866025                        210.0  
2024-12-15 23:00:00    1028.0 -0.939693 -0.342020                        250.0  
2024-12-16 00:00:00    1028.3 -0.965926 -0.258819                        255.0  
2024-12-16 01:00:00    1028.0 -0.984808 -0.173648                        260.0  
2024-12-16 02:00:00    1028.0 -0.984808  0.173648                        280.0  
2024-12-16 03:00:00    1027.0 -0.939693 -0.342020                        250.0  
2024-12-16 04:00:00    1026.6 -0.927184 -0.374607                        248.0  
2024-12-16 05:00:00    1025.7 -0.970296 -0.241922                        256.0  
2024-12-16 06:00:00    1024.8 -0.990268 -0.139173                        262.0  
2024-12-16 07:00:00    1024.8 -0.994522 -0.104528                        264.0  
2024-12-16 08:00:00    1024.9 -0.998630 -0.052336                        267.0  
2024-12-16 09:00:00    1024.9 -0.999848  0.017452                        271.0  
2024-12-16 10:00:00    1025.2 -0.945519 -0.325568                        251.0  
2024-12-16 11:00:00    1025.5 -0.754710 -0.656059                        229.0  
2024-12-16 12:00:00    1025.7 -0.500000 -0.866025                        210.0  
2024-12-16 13:00:00    1025.9 -0.515038 -0.857167                        211.0  
2024-12-16 14:00:00    1026.0 -0.544639 -0.838671                        213.0  
2024-12-16 15:00:00    1026.1 -0.573576 -0.819152                        215.0

## wind direction is in degrees so directly interpolating them will not take into account the circular data
### the direction is broken into sine and cosine components 

In [27]:
import numpy as np
data_filtered['wdir_sin'] = np.sin(np.radians(data_filtered['Wind Direction']))
data_filtered['wdir_cos'] = np.cos(np.radians(data_filtered['Wind Direction']))

## linearly interpolate values with small gaps i.e 2

In [28]:
data_filtered['wdir_sin'] = data_filtered['wdir_sin'].interpolate(method='linear', limit=2)
data_filtered['wdir_cos'] = data_filtered['wdir_cos'].interpolate(method='linear', limit=2)

In [29]:
# Recalculate wind direction from interpolated sine and cosine
data_filtered['Wind_Direction_Interpolated'] = np.degrees(np.arctan2(data_filtered['wdir_sin'], data_filtered['wdir_cos']))

# Ensure wind direction is between 0 and 360 degrees
data_filtered['Wind_Direction_Interpolated'] = data_filtered['Wind_Direction_Interpolated'] % 360
# Replace original Wind_Direction with interpolated values where Wind_Direction was NaN
data_filtered['Wind Direction'] = data_filtered['Wind Direction'].fillna(data_filtered['Wind_Direction_Interpolated'])


## only 483 null values remain

In [30]:
data_filtered.isnull().sum()

Timestamp                        0
Temperature                      0
Wind Direction                 483
Wind Speed                       0
Humidity                         0
Pressure                         0
wdir_sin                       483
wdir_cos                       483
Wind_Direction_Interpolated    483
dtype: int64

In [31]:
data_filtered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96064 entries, 0 to 96063
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Timestamp                    96064 non-null  datetime64[ns]
 1   Temperature                  96064 non-null  float64       
 2   Wind Direction               95581 non-null  float64       
 3   Wind Speed                   96064 non-null  float64       
 4   Humidity                     96064 non-null  float64       
 5   Pressure                     96064 non-null  float64       
 6   wdir_sin                     95581 non-null  float64       
 7   wdir_cos                     95581 non-null  float64       
 8   Wind_Direction_Interpolated  95581 non-null  float64       
dtypes: datetime64[ns](1), float64(8)
memory usage: 6.6 MB


## Out of 95053 rows 483 null values remaining with 182 instances having wind speed equal to 0

In [32]:
wind_zero_null_direction=data_filtered[(data_filtered['Wind Direction'].isnull()) & (data_filtered['Wind Speed']==0 )]
wind_zero_null_direction.describe()

Timestamp  Temperature  Wind Direction  Wind Speed  \
count                            182   182.000000             0.0       182.0   
mean   2016-04-10 17:30:59.340659200    10.796703             NaN         0.0   
min              2014-01-01 15:00:00    -4.000000             NaN         0.0   
25%              2015-06-25 03:30:00     4.000000             NaN         0.0   
50%              2016-01-08 19:30:00    10.000000             NaN         0.0   
75%              2017-09-28 02:00:00    16.000000             NaN         0.0   
max              2018-11-28 21:00:00    32.000000             NaN         0.0   
std                              NaN     7.641656             NaN         0.0   

         Humidity     Pressure  wdir_sin  wdir_cos  \
count  182.000000   182.000000       0.0       0.0   
mean    81.802198  1021.523316       NaN       NaN   
min     42.000000  1005.000000       NaN       NaN   
25%     75.250000  1019.000000       NaN       NaN   
50%     82.000000  1022.000000       NaN       NaN   
75%     88.000000  1026.000000       NaN       NaN   
max    100.000000  1036.000000       NaN       NaN   
std     10.877235     6.705822       NaN       NaN   

       Wind_Direction_Interpolated  
count                          0.0  
mean                           NaN  
min                            NaN  
25%                            NaN  
50%                            NaN  
75%                            NaN  
max                            NaN  
std                            NaN

## Aggregating hourly data to monthly data 
### remaining null values in wind direction indicate low wind speed or no wind so we keep them and impute mean of wind direction for that specific month 
### using mean for all other parameters.

In [34]:
data_filtered.set_index('Timestamp', inplace=True)

monthly_data = data_filtered.resample('M').agg({
    'Wind Direction': lambda x: np.degrees(np.arctan2(np.nanmean(np.sin(np.radians(x))), np.nanmean(np.cos(np.radians(x))))) % 360,
    'Wind Speed': 'mean',  # Use mean or other method as preferred
    'Temperature': 'mean',
    'Humidity': 'mean',
    'Pressure': 'mean'
}).where(data_filtered.resample('M').count() >= 20)  # Adjust min_periods as needed


## no null values

In [35]:
monthly_data.isnull().sum()

Wind Direction    0
Wind Speed        0
Temperature       0
Humidity          0
Pressure          0
dtype: int64

In [45]:
monthly_data.tail(15)

Wind Direction  Wind Speed  Temperature   Humidity     Pressure
Timestamp                                                                  
2023-10-31       48.550879   11.081720    21.066801  67.419355  1019.397715
2023-11-30      321.215866   13.935556    14.887500  60.848611  1023.186250
2023-12-31      329.751841   12.502151     7.537634  65.294355  1026.823253
2024-01-31      356.159741   12.846774     6.594086  66.689516  1026.935215
2024-02-29      347.379127   15.516667     6.846264  75.396552  1024.964511
2024-03-31       96.409317   13.397849    12.271909  67.778226  1018.720565
2024-04-30       96.652195   13.155000    17.545833  73.590278  1012.629028
2024-05-31      124.164003   14.695968    21.677419  64.368280  1013.188710
2024-06-30      164.679932   12.019444    24.929167  77.311111  1007.366111
2024-07-31      178.374792   17.010753    31.580645  69.548387  1004.698522
2024-08-31      135.622965   12.851344    31.934140  67.279570  1005.059812
2024-09-30       84.487889   15.679167    28.128889  76.018056  1009.637222
2024-10-31       18.242554   14.990054    20.412634  74.865591  1019.106183
2024-11-30       14.797232   14.293194    15.417639  67.195833  1023.228750
2024-12-31       11.937579   10.410638     9.045213  60.630319  1025.261436

## AIR quality data extracted from a website with monthly data

In [ ]:
air_qa=pd.read_excel('C:/Users/Administrator/Documents/waleed/air_quality_analysis/air_quality.xlsx')

In [ ]:
air_qa.head()

In [ ]:
air_qa.head()

In [ ]:
air_qa.describe()

## no null values in air quality

In [ ]:
air_qa.info()

## need to combine air quality data with the dataframe for other weather conditions temperature,pressure etc

# converting air quality date column 'month' to datetime

In [ ]:
air_qa['month']=pd.to_datetime(air_qa['month'])

In [ ]:
air_qa.head(50)

In [ ]:
monthly_data.reset_index(inplace=True)

In [ ]:
monthly_data.head(50)

## there are days in date column of both tables so removing them to join them on month and year

In [ ]:
monthly_data['Year'] = monthly_data['Timestamp'].dt.year
monthly_data['Month'] = monthly_data['Timestamp'].dt.month
air_qa['Year'] = air_qa['month'].dt.year
air_qa['Month'] = air_qa['month'].dt.month



## finally merging the data based on month and year to get complete dataset 

In [ ]:
combined_df = pd.merge(monthly_data, air_qa, on=['Year', 'Month'], how='inner')

In [ ]:
combined_df = combined_df.drop(columns=['Timestamp', 'month'])

## Final data

In [ ]:
combined_df.tail(50)

## converting to excel file

In [ ]:
file_path = f'C:/Users/Administrator/Documents/waleed/air_quality_analysis/combined_data.xlsx'  # You can provide a full path if needed

# Save the DataFrame to an Excel file
combined_df.to_excel(file_path, index=False)

print(f"DataFrame saved to {file_path}")